godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,818 @@
+        p_instance: sys::GDExtensionScriptInstanceDataPtr,
+        p_prop_info: *const sys::GDExtensionPropertyInfo,
+    ) {
+        let ctx = || format!("error while calling {}::property_list", type_name::<T>());
+
+        let length = handle_panic(ctx, || {
+            let instance = instance_data_as_script_instance::<T>(p_instance);
+
+            borrow_instance(instance).property_list().len()
+        })
+        .unwrap_or_default();
+
+        // SAFETY: p_prop_info is expected to have been created by get_property_list_func
+        // and therefore should have the same length as before. get_propery_list_func
+        // also guarantees that both vector length and capacity are equal.
+        let vec = transfer_ptr_list_from_godot(p_prop_info, length);
+
+        drop(vec);
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid pointer to the instance data.
+    /// - `p_method` has to be a valid Godot string name ptr.
+    /// - `p_args` has to point to a list of Variant pointers of length `p_argument_count`.
+    /// - `r_return` has to be a valid variant pointer into which the return value can be moved.
+    /// - `r_error` has to point to a valid [`GDExtenstionCallError`] which can be modified to reflect the outcome of the method call.
+    pub(super) unsafe extern "C" fn call_func<T: ScriptInstance>(
+        p_self: sys::GDExtensionScriptInstanceDataPtr,
+        p_method: sys::GDExtensionConstStringNamePtr,
+        p_args: *const sys::GDExtensionConstVariantPtr,
+        p_argument_count: sys::GDExtensionInt,
+        r_return: sys::GDExtensionVariantPtr,
+        r_error: *mut sys::GDExtensionCallError,
+    ) {
+        let method = transfer_string_name_from_godot(p_method);
+        let args = Variant::unbounded_refs_from_sys(p_args, p_argument_count as usize);
+        let ctx = || format!("error when calling {}::call", type_name::<T>());
+
+        let result = handle_panic(ctx, || {
+            let instance = instance_data_as_script_instance::<T>(p_self);
+            borrow_instance_mut(instance).call(method.clone(), args)
+        });
+
+        match result {
+            Some(Ok(ret)) => {
+                transfer_variant_to_godot(ret, r_return);
+                (*r_error).error = sys::GDEXTENSION_CALL_OK;
+            }
+
+            Some(Err(err)) => (*r_error).error = err,
+            None => {
+                (*r_error).error = sys::GDEXTENSION_CALL_ERROR_INVALID_METHOD;
+            }
+        };
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid pointer to the instance data.
+    /// - A pointer to a `Script` reference will be returned. The caller is then responsible for freeing the reference, including the adjustment
+    ///   of the reference count.
+    pub(super) unsafe extern "C" fn get_script_func<T: ScriptInstance>(
+        p_instance: sys::GDExtensionScriptInstanceDataPtr,
+    ) -> sys::GDExtensionObjectPtr {
```

Comment:
The engine actually takes this pointer and constructs its own `Ref<Script>`. This is why the previous revision of this PR caused the reference count to be off and the `Script` to be leaked. 

The current implementation of `transfer_script_to_godot` works as long as the reference count of `Script` is > 1. Otherwise, the count drops to 0 before we return and the `ptr` is no longer valid.

If we make `ScriptInstance` return a `&Gd<Script>` we could guarantee that it outlives T and therefore the reference count won't go to 0 before the engine constructs it's ´Ref<T>`. @Bromeon what do you think?

Reply:
> [`Script`](https://docs.godotengine.org/en/stable/classes/class_script.html) itself is reference-counted, so how would `ScriptInstance::get_script` returning `&Gd<Script>` instead of `Gd<Script>` ensure that it outlives `T`?

Someone could implement `get_script` as
```rs
fn get_script(&self) -> Gd<Script> {
  Gd::<MyScritp>::new()
}
```

Reply:
@lilizoey right that was what I thought as well.

Reply:
get_script now returns a reference.